Automattic/woocommerce-payments
# Convert the fraud protection settings files to TypeScript

Fixes #6031 

#### Changes proposed in this Pull Request



This pull request changes all the JavaScript files to TypeScript files in the `client/settings/fraud-protection` directory and its subdirectories.







#### Testing instructions





- Checkout PR branch.
- Build the client: `npm run build:client`
- All tests should pass.
- Make sure the feature flag option is set to `1`: `npm run wp update option wcpay_fraud_protection_settings_active 1`
- Visit the WCPay settings page and scroll to the bottom. The fraud protection section should look like the following screenshot:

![pr6032_wcpay-settings-fraud-protection](https://user-images.githubusercontent.com/60988591/233705340-228a5ac9-88c5-413f-8536-c9a7820c7f77.png)

- Click the question mark icon next to Basic. A modal should appear and look like the following screenshot:

![pr6032_basic-protection-modal png](https://user-images.githubusercontent.com/60988591/233705418-57fca7a8-ed76-41d5-8e0a-f00e12c3bf0e.png)

- Perform each of the following actions on the modal to confirm that it closes with each action:
  - Click the 'Got it' button.
  - Click the 'X' button.
  - Click anywhere outside of the modal.

- Click the advanced radio button, then the configure button.
- Check that each of the following advanced rule cards are rendered on the page:
  - AVS Mismatch
  - CVC Verification
  - International IP Address
  - IP Address Mismatch
  - Address Mismatch
  - Purchase Price Threshold
  - Order Items Threshold

- Make changes to each rule ensuring that the toggle works for each rule. Set each rule to only hold for review. For purchase price threshold rule set minimum to 10 and max to 100. For order items threshold rule set minimum to 2, max to 20. Save changes.
- Inspect the saved ruleset in your local client and local server DBs to ensure that the rules were saved correctly.

Client DB:
 
```sql
SELECT * FROM `wp_options` WHERE `option_name` = '_transient_wcpay_fraud_protection_settings';
```

Server DB:

```sql
SELECT * FROM `wcpay_fraud_rulesets` ORDER BY `wcpay_fraud_rulesets`.`wcpay_fraud_ruleset_id` DESC;
```

- Now set each of the rules to block. Save changes.
- Inspect the saved ruleset in your local client and local server DBs to ensure that the rules were saved correctly.
- Now remove the purchase price threshold and order items threshold minimums. Save settings.
- Inspect the saved ruleset to ensure that the purchase price threshold and order items threshold rules were saved correctly.
- Now re-add the purchase price threshold and order items threshold minimums. Remove the purchase price threshold and order items threshold maximums. 
  - You should receive the following error notice on the order items threshold when there is a minimum value but no max value: _Maximum item count must be greater than the minimum item count._ Delete the minumum value, you should get the following notice: _An item range must be set for this filter to take effect._ 
  - Toggle the order items threshold filter off. 
  - Save settings.
- Inspect the saved ruleset to ensure that the purchase price threshold and order items threshold rules were saved correctly.
- Test the same notices on the purchase price threshold rule:
  - Delete both the min and max values, you should get the following notice: _A price range must be set for this filter to take effect._
  - Put a value in the minimum field, then put a smaller value in the max field. You should get the following error notice: _Maximum purchase price must be greater than the minimum purchase price._
- Now set the following advanced config:
  - Disable filtering for: International IP Address, IP Address Mismatch, and Address Mismatch.
  - Purchase price threshold: leave min field empty, set max field to 50, uncheck the block payment box.
  - Order items threshold: leave min field empty, set max field to 10, check the block payment box.
  - Save settings.
- Check that a new `wcpay_fraud_rule_id` was created for this new saved ruleset on the server DB.
- Now click save settings again. 
- Check that a new `wcpay_fraud_rule_id` was ***NOT*** created for this new saved ruleset on the server DB.
- Now, go back to the WCPay settings page, set your fraud protection level to basic and save settings. Check that the fraud ruleset was correctly saved on the client DB and server DB.
- Lastly, click save settings again and check that a new `wcpay_fraud_rule_id` was ***NOT*** created for this new saved ruleset on the server DB.

-------------------


**Post merge**
